Dev

ksabbak opened the pull request.

ksabbak committed:
Allow players to choose marker, stop using atoms

Players can now choose their markers and the marker of the computer opponent. The marker must be one character in length and cannot be the same as the other marker.

Player status is now being determined by a "turn" instead of clojure's atom data structure for a more functional approach.

ksabbak committed:
Fix formatting, styles, etc.

Mostly in response to comments on PR 15
https://github.com/ksabbak/ClojureTTT/pull/15

Some in response to comments in PR 14
https://github.com/ksabbak/ClojureTTT/pull/14

ksabbak committed:
Merge pull request #15 from ksabbak/players

Allow players to choose marker, stop using atoms

ksabbak committed:
Allow for game-type choice

Game can now be played human vs. human or human vs. computer.

ksabbak committed:
Fix function name, hard-coding

In response to comments on  https://github.com/ksabbak/ClojureTTT/pull/16

Changes to players 'hardcoding' is more of a band-aid than a permanent fix

ksabbak committed:
Merge pull request #16 from ksabbak/game-type

Allow for game-type choice

ksabbak committed:
Clean indendation, parens

BWoodfork commented on src/clojure_tictactoe/cli/input/input_getter.clj:
```diff
@@ -13,25 +14,38 @@
-        formatted-input))))
+(defn parse-move-input [input choice-limit]
+  (let [formatted-input (read-string input)
+        valid-numeral (and
```
Can this `let` binding for `valid-numeral` be more descriptive?

ksabbak replied:
Do you have a suggestion? Because I'm actually rather proud of that name.

BWoodfork commented on src/clojure_tictactoe/game/game_play.clj:
```diff
@@ -8,27 +8,36 @@
-  (board/mark-space (move-function board) @players/player-atom board))
+(def standard-board-size 9)
+(def game-options ["Human vs Human" "Human vs. Computer"])
+(def players '("player 1" "player 2"))
```
Is there a reason these are used as `def`'s and not `let` bindings?

ksabbak replied:
I guess there is not?

BWoodfork commented on src/clojure_tictactoe/cli/input/input_getter.clj:
```diff
@@ -13,25 +14,38 @@
+    (println instructions-printer/game-choice-request)
+    (let [choice-limit (+ (count options) 1)]
+      (if-let [choice (parse-move-input (get-user-input) choice-limit)]
+        (options (- choice 1))
```
what is happening here?

ksabbak replied:
parse-move-input will be renamed later, but essentially what's happening is the input is looking for a valid choice and then returning the choice (rather than the corresponding number so if the order gets changed nothing else has to be changed).

BWoodfork commented on src/clojure_tictactoe/game/game_rules.clj:
```diff
@@ -8,10 +8,10 @@
-      (list top-left-to-bottom-right-diagonal bottom-left-to-top-right-diagonal))))
+  (let [side-length (board/side-length board)
+        top-left-to-bottom-right-diagonal (take side-length (range 0 (count board) (+ 1 side-length)))
+        bottom-left-to-top-right-diagonal (take side-length (range (- side-length 1) (count board) (- side-length 1)))]
```
I think these can just be called "left-diagonal" and "right-diagonal"

ksabbak replied:
Probably, but I couldn't remember which one was which when I did that, so this is more for my benefit.

BWoodfork replied:
Does it matter which is which?

ksabbak replied:
I guess not.

BWoodfork commented on src/clojure_tictactoe/game/game_rules.clj:
```diff
@@ -8,10 +8,10 @@
+  (let [side-length (board/side-length board)
+        top-left-to-bottom-right-diagonal (take side-length (range 0 (count board) (+ 1 side-length)))
+        bottom-left-to-top-right-diagonal (take side-length (range (- side-length 1) (count board) (- side-length 1)))]
+    (list top-left-to-bottom-right-diagonal bottom-left-to-top-right-diagonal)))
```
See - https://stackoverflow.com/questions/1147975/in-clojure-when-should-i-use-a-vector-over-a-list-and-the-other-way-around

ksabbak replied:
Since I'm never accessing them outside of the order, does this matter? I am perfectly willing to change it anyway.

BWoodfork commented on src/clojure_tictactoe/game/game_rules.clj:
```diff
@@ -23,7 +23,9 @@
 
 (defn assess-winner [board]
-  (first (some #(when (apply = %) %) (potential-wins board))))
+  (-> #(when (apply = %) %)
```
Sometimes anonymous functions can be pretty nice. But sometimes, they can make code a little more difficult to read. This is one of those times lol. I would name the arg in a function here instead of using an anonymous function.

ksabbak replied:
✅

BWoodfork commented on src/clojure_tictactoe/game/players.clj:
```diff
@@ -1,21 +1,40 @@
+  (let [markers (map acquire-one-marker players)]
+    (if (distinct-markers? markers)
+      (into [] markers)
+      (recur players))))
```
Why is this function recurring? I'm assuming because it's waiting to receive two valid markers. If that's the case, then this is doing more than one thing. Whatever is responsible for handling the validation/user input, should be the thing to call recur (you input getter). This namespace shouldn't be responsible for whether or not both markers were received.

ksabbak replied:
✅

hnlee commented on src/clojure_tictactoe/cli/input/input_getter.clj:
```diff
@@ -13,25 +14,38 @@
-            (recur))))))
+    (println instructions-printer/game-choice-request)
+    (let [choice-limit (+ (count options) 1)]
+      (if-let [choice (parse-move-input (get-user-input) choice-limit)]
```
I thought I left a comment about this on your last PR, but I think you really need to rename `parse-move-input` because here, it is definitely not parsing move input.

ksabbak replied:
You did. And I agree.

ksabbak committed:
Refactor to pull validation out of input-getter

In order to better follow the SRP and in response to comments on https://github.com/ksabbak/ClojureTTT/pull/17

Smaller changes have also been made in response to those comments.

Display has been changed for the board to go 1-9 instead of 0-8 because it allows for smoother checking and also a lot of people (esp. non-programmers) have a hard time with the idea of zero-first numbering systems.

BWoodfork commented on src/clojure_tictactoe/cli/input/input_getter.clj:
```diff
@@ -13,25 +14,53 @@
+(defn format-input [input]
+  (if (or
+        (empty? input)
+        (string/includes? input "#"))
```
Can this `or` condition be expressed more explicitly?

ksabbak replied:
Can you elaborate what you're looking for?

BWoodfork commented on src/clojure_tictactoe/cli/output/board_printer.clj:
```diff
@@ -1,19 +1,24 @@
 (ns clojure-tictactoe.cli.output.board-printer
   (:require [clojure-tictactoe.game.board :as board]))
 
+(defn space-renderer [space]
```
Does this function render spaces?

ksabbak replied:
`space-content-renderer`?

BWoodfork commented on src/clojure_tictactoe/game/game_play.clj:
```diff
@@ -8,27 +8,35 @@
+                      game-type turn)
+                    board
+                    (markers
+                      (players/current-player turn)))]
```
The things happening within this let variable are a little crazy and difficult to read. Can you clean this up a bit?

ksabbak replied:
✅

BWoodfork commented on src/clojure_tictactoe/game/game_rules.clj:
```diff
@@ -22,8 +22,12 @@
   (some? (some #(apply = %) (potential-wins board))))
 
+
+(defn check-for-equality [potential-wins board]
```
what's considered "equality"?

ksabbak replied:
Hah, no idea what happened there or why that didn't break the tests when I ran them the first time. Fixed it to what it should have been.

BWoodfork commented on src/clojure_tictactoe/cli/input/input_checker.clj:
```diff
@@ -0,0 +1,25 @@
+
+(defn valid-move? [user-input board]
+  (and
+    (board/space-is-open? user-input board)))
```
`and` is being used for?

ksabbak replied:
Something that's not there anymore.

hnlee commented on src/clojure_tictactoe/cli/input/input_getter.clj:
```diff
@@ -1,7 +1,7 @@
-            [clojure-tictactoe.game.board :as board]
-            [clojure-tictactoe.cli.output.instructions-printer :as instructions-printer]))
+            [clojure-tictactoe.cli.output.instructions-printer :as instructions-printer]
+            [clojure-tictactoe.cli.input.input-checker :as input-checker]))
```
I like these namespace changes!

ksabbak replied:
Thanks

ksabbak committed:
Fix check-for-equality, clean up.

check-for-equality didn't do what it needed to, it's now assess-potential-wins and it works
general clean up and renaming

ksabbak committed:
Rename vars, clean spacing
